From 66c4625aaff6d8bc1b88f1001b753e2b79712949 Mon Sep 17 00:00:00 2001
From: Christian Bruel <christian.bruel@foss.st.com>
Date: Wed, 6 Sep 2023 15:10:03 +0200
Subject: [PATCH 0502/1141] PCI: pci-epf-test: Add force_core_init flag in
 epc_features

Allow device initialisation to be done also when binding regardless
of being notified later when starting the link.

Signed-off-by: Christian Bruel <christian.bruel@foss.st.com>
Change-Id: Ia0a66ef6a3f26f0dbd587362d7188ac616e4dac9
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/327478
Reviewed-by: Fabrice GASNIER <fabrice.gasnier@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Fabrice GASNIER <fabrice.gasnier@foss.st.com>
---
 drivers/pci/endpoint/functions/pci-epf-test.c | 4 +++-
 include/linux/pci-epc.h                       | 2 ++
 2 files changed, 5 insertions(+), 1 deletion(-)

diff --git a/drivers/pci/endpoint/functions/pci-epf-test.c b/drivers/pci/endpoint/functions/pci-epf-test.c
index 2e8a4de2abab..0cd96ce0b846 100644
--- a/drivers/pci/endpoint/functions/pci-epf-test.c
+++ b/drivers/pci/endpoint/functions/pci-epf-test.c
@@ -959,6 +959,7 @@ static int pci_epf_test_bind(struct pci_epf *epf)
 	struct pci_epc *epc = epf->epc;
 	bool linkup_notifier = false;
 	bool core_init_notifier = false;
+	bool force_core_init = false;
 
 	if (WARN_ON_ONCE(!epc))
 		return -EINVAL;
@@ -971,6 +972,7 @@ static int pci_epf_test_bind(struct pci_epf *epf)
 
 	linkup_notifier = epc_features->linkup_notifier;
 	core_init_notifier = epc_features->core_init_notifier;
+	force_core_init = epc_features->force_core_init;
 	test_reg_bar = pci_epc_get_first_free_bar(epc_features);
 	if (test_reg_bar < 0)
 		return -EINVAL;
@@ -983,7 +985,7 @@ static int pci_epf_test_bind(struct pci_epf *epf)
 	if (ret)
 		return ret;
 
-	if (!core_init_notifier) {
+	if (!core_init_notifier || force_core_init) {
 		ret = pci_epf_test_core_init(epf);
 		if (ret)
 			return ret;
diff --git a/include/linux/pci-epc.h b/include/linux/pci-epc.h
index a48778e1a4ee..ebf54a619cef 100644
--- a/include/linux/pci-epc.h
+++ b/include/linux/pci-epc.h
@@ -154,6 +154,7 @@ struct pci_epc {
 
 /**
  * struct pci_epc_features - features supported by a EPC device per function
+ * @force_core_init: indicate if the EPC device can initialize when binding
  * @linkup_notifier: indicate if the EPC device can notify EPF driver on link up
  * @core_init_notifier: indicate cores that can notify about their availability
  *			for initialization
@@ -165,6 +166,7 @@ struct pci_epc {
  * @align: alignment size required for BAR buffer allocation
  */
 struct pci_epc_features {
+	unsigned int	force_core_init : 1;
 	unsigned int	linkup_notifier : 1;
 	unsigned int	core_init_notifier : 1;
 	unsigned int	msi_capable : 1;
-- 
2.39.2

